### ISSUE TITLE
Input on structure for models slot in row data featuresClass

### ISSUE THREAD

**Issue by statOmics** [statOmics/msqrob2#4]
Stuck with structure

Low level methods for extracting contrasts, unscaled covariance matrix, degrees of freedom and residual variance are available.

Can make a new S4 vector class of msqrobModel class objects e.g. 
setClass(Class = "msqrobModels", 
    representation = list(x = "msqrobModel"), 
    contains = "Vector")

Current msqrob function to fit the models is using an apply loop that returns an ordinary list. 
How can I recast that list into an the new S4 vector class or is there an alternative to apply to return a Vector class?

**Comment by lgatto** [comment]
@ococrook - you had something yesterday, didn't you.

I would suggest something more generic, like
```r
.StatModel <- setClass("StatModel",  slots = c(type = "character", params = "list"))
```
It would then used as such:
```r
> mod1 <- .StatModel(type = "msqrob")
> mod2 <- .StatModel(type = "limma")
> df <- DataFrame(x = 1:2)
> df$mods <- c(mod1, mod2)
> df
DataFrame with 2 rows and 2 columns
          x                                               mods
  <integer>                                             <list>
1         1 new("StatModel", type = "msqrob", params = list())
2         2  new("StatModel", type = "limma", params = list())
```
These could also be sub-classes with default `type` slots. In the case of the *msqrob* type, `params` would get the list you mention above.

**Comment by statOmics** [comment]
In the current implementation of msqrob function that mimics limma but will handle the NAs appropriately, I already return a list of msqrobModels. These can be added to the row data. However, upon fitting I need to add the posterior variances and posterior degrees of freedom. I could add that as a next column to the rowData, however, I would prefer to add it to the list of Models. I could provide an empty slot to the msqrobModels class to add the posterior variances later. However, accessing the msqrobModels or Statmodels in the list to fill additional slots is probably not efficient, is it?

**Comment by lgatto** [comment]
Yes, add these to the models. 

I don't think we need to worry about efficiency at this stage.

**Comment by statOmics** [comment]
Code works for ols and M-estimation. 
Perhaps good to evaluate the solution for posterior variance and degrees of freedom. (see msqrob function)

Vignette with cptac 3 vs 3 is also working with all plots

**Comment by ococrook** [comment]
@statOmics @lgatto 

I already wrote the following. The param slot can be extremely general, since it is just a list. I can also add a slot for "type" - I'll update the show method accordingly, as well as Laurent's examples. Did you also want a slot for variances or is to be included inside the param list?

@statOmics I also think the "representation" way to create a class slot is now best avoided, ```slots``` should do the job

```{r,}
##' classes for msqrob
##'
##' @title The StatModel class for msqrob
##' @slot param A list containing information of the used model
##' @rd
##' @rdname msqrob-framework
##' @author Oliver M. Crook
##' @examples 
##' myModel <- new("StatModel", param = list(x = 3, y = 7, b = 4))
##' myModel@param
setClass("StatModel",
         slots = c(param = "list"))

##' @title show method for object of class StatModel
##' 
##' @rdname msqrob-framework
setMethod("show", "StatModel",
          function(object) {
            cat("Object of class \"", class(object), "\"\n", sep = "")
            cat("There number of elements is", length(object@param),"\n")
            invisible(NULL)
          })

````

**Comment by statOmics** [comment]
Excellent, 

@ococrook Just go ahead. 

The msqrobModel class was in fact the blue print for the StatModel class. It also has a type and a list slot. I used msqrob in the name so as to make it clear that it is an object belonging to the msqrob package.  

I would opt for var.post and a df.post slot. 
I included that in a version that I did not push to GitHub, but I wanted to avoid to have to loop over the list of StatModel or msqrobModel elements to assign a vector of the posterior variances. 
So the msqrob function temporarily returns a DataFrame with a column with the posterior variances, degrees of freedom and the models. 
Is there an efficient way to assign each element of a vector to a specific slot of an S4 object element in a list? 

Thanks also for the update related to the representation way. Is a long time ago that I initiated a package myself...
